[Bug]: Cannot find module '$SNIP/packages/storybook/package.json'

### Describe the bug

Upgrading from 7.0.23 -> 7.0.24

```bash
ERR! Error: Cannot find module '$SNIP/packages/storybook/package.json'
ERR! Require stack:
ERR! - /home/circleci/project/node_modules/@storybook/telemetry/dist/index.js
ERR! - /home/circleci/project/node_modules/@storybook/cli/dist/generate.js
ERR! - /home/circleci/project/node_modules/@storybook/cli/bin/index.js
ERR! - /home/circleci/project/node_modules/storybook/index.js
ERR!     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
ERR!     at Module._resolveFilename (/home/circleci/project/node_modules/esbuild-register/dist/node.js:4768:36)
ERR!     at Function.resolve (node:internal/modules/cjs/helpers:116:19)
ERR!     at getActualPackageJson (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:2095)
ERR!     at getFrameworkInfo (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:5518)
ERR!     at computeStorybookMetadata (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:7772)
ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ERR!     at async getStorybookMetadata (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:9731)
ERR!     at async extractStorybookMetadata (/home/circleci/project/node_modules/@storybook/core-server/dist/index.js:20:6098)
ERR!     at async Promise.all (index 5)
ERR!  Error: Cannot find module '$SNIP/packages/storybook/package.json'
ERR! Require stack:
ERR! - /home/circleci/project/node_modules/@storybook/telemetry/dist/index.js
ERR! - /home/circleci/project/node_modules/@storybook/cli/dist/generate.js
ERR! - /home/circleci/project/node_modules/@storybook/cli/bin/index.js
ERR! - /home/circleci/project/node_modules/storybook/index.js
ERR!     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
ERR!     at Module._resolveFilename (/home/circleci/project/node_modules/esbuild-register/dist/node.js:4768:36)
ERR!     at Function.resolve (node:internal/modules/cjs/helpers:116:19)
ERR!     at getActualPackageJson (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:2095)
ERR!     at getFrameworkInfo (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:5518)
ERR!     at computeStorybookMetadata (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:7772)
ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ERR!     at async getStorybookMetadata (/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js:1:9731)
ERR!     at async extractStorybookMetadata (/home/circleci/project/node_modules/@storybook/core-server/dist/index.js:20:6098)
ERR!     at async Promise.all (index 5) {
ERR!   code: 'MODULE_NOT_FOUND',
ERR!   requireStack: [
ERR!     '/home/circleci/project/node_modules/@storybook/telemetry/dist/index.js',
ERR!     '/home/circleci/project/node_modules/@storybook/cli/dist/generate.js',
ERR!     '/home/circleci/project/node_modules/@storybook/cli/bin/index.js',
ERR!     '/home/circleci/project/node_modules/storybook/index.js'
ERR!   ]
ERR! }
```

### To Reproduce

https://github.com/ntucker/anansi/pull/2061

### System

```shell
Environment Info:

  System:
    OS: Linux 5.15 Ubuntu 20.04 LTS (Focal Fossa)
    CPU: (20) x64 AMD Ryzen 9 7950X 16-Core Processor
  Binaries:
    Node: 20.3.1 - ~/.nvm/versions/node/v20.3.1/bin/node
    Yarn: 3.5.0 - ~/.yarn/bin/yarn
    npm: 9.6.7 - ~/.nvm/versions/node/v20.3.1/bin/npm
```


### Additional context

storybook with react and webpack

https://app.circleci.com/pipelines/github/ntucker/anansi/9350/workflows/c40db23e-7d88-4bf1-9f36-103b8bcc2245/jobs/50149

--- a/code/lib/telemetry/src/get-framework-info.test.ts
+++ b/code/lib/telemetry/src/get-framework-info.test.ts
@@ -1,0 +1,55 @@
+import type { StorybookConfig } from '@storybook/types';
+import { getFrameworkInfo } from './get-framework-info';
+import { getActualPackageJson } from './package-json';
+
+jest.mock('./package-json', () => ({
+  getActualPackageJson: jest.fn(),
+}));
+
+describe('getFrameworkInfo', () => {
+  it('should return an empty object if mainConfig.framework is undefined', async () => {
+    const result = await getFrameworkInfo({} as StorybookConfig);
+    expect(result).toEqual({});
+  });
+
+  it('should return an empty object if mainConfig.framework name is undefined', async () => {
+    const result = await getFrameworkInfo({ framework: {} } as StorybookConfig);
+    expect(result).toEqual({});
+  });
+
+  it('should call getActualPackageJson with the correct package name', async () => {
+    const packageName = '@storybook/react';
+    const framework = { name: packageName };
+    await getFrameworkInfo({ framework } as StorybookConfig);
+    expect(getActualPackageJson).toHaveBeenCalledWith(packageName);
+  });
+
+  it('should resolve the framework package json correctly and strip project paths in the metadata', async () => {
+    const packageName = '/path/to/project/@storybook/react';
+    const framework = { name: packageName };
+    const frameworkPackageJson = {
+      name: packageName,
+      dependencies: {
+        '@storybook/react': '7.0.0',
+        '@storybook/builder-vite': '7.0.0',
+      },
+    };
+
+    jest.spyOn(process, 'cwd').mockReturnValue('/path/to/project');
+
+    (getActualPackageJson as jest.Mock).mockResolvedValueOnce(frameworkPackageJson);
+
+    const result = await getFrameworkInfo({ framework } as StorybookConfig);
+
+    expect(getActualPackageJson).toHaveBeenCalledWith(packageName);
+
+    expect(result).toEqual({
+      framework: {
+        name: '$SNIP/@storybook/react',
+        options: undefined,
+      },
+      builder: '@storybook/builder-vite',
+      renderer: '@storybook/react',
+    });
+  });
+});

--- a/code/lib/telemetry/src/get-framework-info.ts
+++ b/code/lib/telemetry/src/get-framework-info.ts
@@ -49,21 +49,26 @@ export const getFrameworkPackageName = (mainConfig?: StorybookConfig) => {
 };
 
 export async function getFrameworkInfo(mainConfig: StorybookConfig) {
-  if (!mainConfig.framework) return {};
+  if (!mainConfig?.framework) return {};
 
-  const frameworkName = getFrameworkPackageName(mainConfig);
-  if (!frameworkName) return {};
-  const frameworkOptions =
-    typeof mainConfig.framework === 'object' ? mainConfig.framework.options : {};
+  const rawName =
+    typeof mainConfig.framework === 'string' ? mainConfig.framework : mainConfig.framework?.name;
+  if (!rawName) return {};
 
-  const frameworkPackageJson = await getActualPackageJson(frameworkName);
+  const frameworkPackageJson = await getActualPackageJson(rawName);
 
   const builder = findMatchingPackage(frameworkPackageJson, knownBuilders);
   const renderer = findMatchingPackage(frameworkPackageJson, knownRenderers);
 
+  // parse framework name and strip off pnp paths etc.
+  const sanitizedFrameworkName = getFrameworkPackageName(mainConfig);
+
+  const frameworkOptions =
+    typeof mainConfig.framework === 'object' ? mainConfig.framework.options : {};
+
   return {
     framework: {
-      name: frameworkName,
+      name: sanitizedFrameworkName,
       options: frameworkOptions,
     },
     builder,

--- a/code/lib/telemetry/src/get-framework-info.ts
+++ b/code/lib/telemetry/src/get-framework-info.ts
@@ -54,12 +54,15 @@ export async function getFrameworkInfo(mainConfig: StorybookConfig) {
 
   const frameworkPackageJson = await getActualPackageJson(rawName);
 
+  if (!frameworkPackageJson) {
+    return {};
+  }
+
   const builder = findMatchingPackage(frameworkPackageJson, knownBuilders);
   const renderer = findMatchingPackage(frameworkPackageJson, knownRenderers);
 
   // parse framework name and strip off pnp paths etc.
   const sanitizedFrameworkName = getFrameworkPackageName(rawName);
-
   const frameworkOptions =
     typeof mainConfig.framework === 'object' ? mainConfig.framework.options : {};
 

--- a/code/lib/telemetry/src/get-framework-info.ts
+++ b/code/lib/telemetry/src/get-framework-info.ts
@@ -33,14 +33,7 @@ function findMatchingPackage(packageJson: PackageJson, suffixes: string[]) {
   return suffixes.map((suffix) => `@storybook/${suffix}`).find((pkg) => allDependencies[pkg]);
 }
 
-export const getFrameworkPackageName = (mainConfig?: StorybookConfig) => {
-  const packageNameOrPath =
-    typeof mainConfig?.framework === 'string' ? mainConfig.framework : mainConfig?.framework?.name;
-
-  if (!packageNameOrPath) {
-    return null;
-  }
-
+export const getFrameworkPackageName = (packageNameOrPath: string) => {
   const normalizedPath = path.normalize(packageNameOrPath).replace(new RegExp(/\\/, 'g'), '/');
 
   const knownFramework = Object.keys(frameworkPackages).find((pkg) => normalizedPath.endsWith(pkg));
@@ -49,11 +42,15 @@ export const getFrameworkPackageName = (mainConfig?: StorybookConfig) => {
 };
 
 export async function getFrameworkInfo(mainConfig: StorybookConfig) {
-  if (!mainConfig?.framework) return {};
+  if (!mainConfig?.framework) {
+    return {};
+  }
 
   const rawName =
     typeof mainConfig.framework === 'string' ? mainConfig.framework : mainConfig.framework?.name;
-  if (!rawName) return {};
+  if (!rawName) {
+    return {};
+  }
 
   const frameworkPackageJson = await getActualPackageJson(rawName);
 
@@ -61,7 +58,7 @@ export async function getFrameworkInfo(mainConfig: StorybookConfig) {
   const renderer = findMatchingPackage(frameworkPackageJson, knownRenderers);
 
   // parse framework name and strip off pnp paths etc.
-  const sanitizedFrameworkName = getFrameworkPackageName(mainConfig);
+  const sanitizedFrameworkName = getFrameworkPackageName(rawName);
 
   const frameworkOptions =
     typeof mainConfig.framework === 'object' ? mainConfig.framework.options : {};

